src/System Application/App/Azure Function/AzureFunctionsOAuth2Cert.Codeunit.al

```diff
@@ -0,0 +1,76 @@
+// ------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+// ------------------------------------------------------------------------------------------------
+
+namespace System.Azure.Functions;
+
+using System.Utilities;
+using System.Security.Authentication;
+using System.Telemetry;
+
+codeunit 7807 "Azure Functions OAuth2 Cert" implements "Azure Functions Authentication"
+{
+    Access = Internal;
+    InherentEntitlements = X;
+    InherentPermissions = X;
+
+    var
+        [NonDebuggable]
+        AuthenticationCodeGlobal, EndpointGlobal : Text;
+        [NonDebuggable]
+        ClientIdGlobal, OAuthAuthorityUrlGlobal, RedirectURLGlobal, ScopeGlobal : Text;
+        CertGlobal: SecretText;
+        AccessToken: SecretText;
+        Scopes: List of [Text];
+        BearerLbl: Label 'Bearer %1', Comment = '%1 is the access token', Locked = true;
+        FailedToGetTokenErr: Label 'Authorization failed to Azure function: %1', Locked = true;
+        AzureFunctionCategoryLbl: Label 'Connect to Azure Functions', Locked = true;
+
+    [NonDebuggable]
+    procedure Authenticate(var RequestMessage: HttpRequestMessage): Boolean
+    var
+        Uri: Codeunit Uri;
+        OAuth2: Codeunit OAuth2;
+        UriBuilder: Codeunit "Uri Builder";
+        FeatureTelemetry: Codeunit "Feature Telemetry";
+        Headers: HttpHeaders;
+        Dimensions: Dictionary of [Text, Text];
+        IdToken: Text;
+    begin
+        UriBuilder.Init(EndpointGlobal);
+        Scopes.Add(ScopeGlobal);
+
+        OAuth2.AcquireTokensWithCertificate(ClientIdGlobal, CertGlobal, RedirectURLGlobal, OAuthAuthorityUrlGlobal, Scopes, AccessToken, IdToken);
+
+        if AccessToken.IsEmpty() then begin
+            UriBuilder.GetUri(Uri);
+            Dimensions.Add('FunctionHost', Format(Uri.GetHost()));
+            FeatureTelemetry.LogError('0000I75', AzureFunctionCategoryLbl, 'Acquiring token', StrSubstNo(FailedToGetTokenErr, Uri.GetHost()), '', Dimensions);
+            exit(false);
+        end;
+
+        RequestMessage.GetHeaders(Headers);
+        Headers.Remove('Authorization');
+        Headers.Add('Authorization', SecretStrSubstNo(BearerLbl, AccessToken));
+
+        if AuthenticationCodeGlobal <> '' then
```

Comment:
If we use the request header `x-functions-key` instead of the query param then we could replace the AuthenticationCodeGlobal with SecretText and probably remove the NonDebuggable from some places.

https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=python-v2%2Cisolated-process%2Cnodejs-v4%2Cfunctionsv2&pivots=programming-language-csharp#api-key-authorization

Reply:
Good comment, that is something we should address in a separate PR though.